packages/slidev/node/resolver.ts

```diff
@@ -0,0 +1,222 @@
+import { dirname, join, resolve } from 'node:path'
+import * as fs from 'node:fs'
+import process from 'node:process'
+import { fileURLToPath } from 'node:url'
+import { ensurePrefix, slash } from '@antfu/utils'
+import isInstalledGlobally from 'is-installed-globally'
+import { resolveGlobal } from 'resolve-global'
+import { findClosestPkgJsonPath, findDepPkgJsonPath } from 'vitefu'
+import { resolvePath } from 'mlly'
+import globalDirs from 'global-directory'
+import prompts from 'prompts'
+import { parseNi, run } from '@antfu/ni'
+import { underline, yellow } from 'kolorist'
+
+export function toAtFS(path: string) {
+  return `/@fs${ensurePrefix('/', slash(path))}`
+}
+
+/**
+ * Find the actual path of the import. If Slidev is installed globally, it will also search globally.
+ */
+export async function resolveImportPath(importName: string, ensure: true): Promise<string>
+export async function resolveImportPath(importName: string, ensure?: boolean): Promise<string | undefined>
+export async function resolveImportPath(importName: string, ensure = false) {
+  try {
+    return await resolvePath(importName, {
+      url: fileURLToPath(import.meta.url),
+    })
+  }
+  catch { }
+
+  if (isInstalledGlobally) {
+    try {
+      return resolveGlobal(importName)
+    }
+    catch { }
+  }
+
+  if (ensure)
+    throw new Error(`Failed to resolve package "${importName}"`)
+}
+
+/**
+ * Find the root of the package. If Slidev is installed globally, it will also search globally.
+ */
+export async function findPkgRoot(dep: string, parent: string, ensure: true): Promise<string>
+export async function findPkgRoot(dep: string, parent: string, ensure?: boolean): Promise<string | undefined>
+export async function findPkgRoot(dep: string, parent: string, ensure = false) {
+  const pkgJsonPath = await findDepPkgJsonPath(dep, parent)
+  const path = pkgJsonPath ? dirname(pkgJsonPath) : isInstalledGlobally ? findGlobalPkgRoot(dep, false) : undefined
+  if (ensure && !path)
+    throw new Error(`Failed to resolve package "${dep}"`)
+  return path
+}
+
+export async function findGlobalPkgRoot(name: string, ensure: true): Promise<string>
+export async function findGlobalPkgRoot(name: string, ensure?: boolean): Promise<string | undefined>
+export async function findGlobalPkgRoot(name: string, ensure = false) {
+  const yarnPath = join(globalDirs.yarn.packages, name)
+  if (fs.existsSync(`${yarnPath}/package.json`))
+    return yarnPath
+  const npmPath = join(globalDirs.npm.packages, name)
+  if (fs.existsSync(`${npmPath}/package.json`))
+    return npmPath
+  if (ensure)
+    throw new Error(`Failed to resolve global package "${name}"`)
+}
+
+export const cliRoot = fileURLToPath(new URL('..', import.meta.url))
+export const clientRoot = await findPkgRoot('@slidev/client', cliRoot, true)
+export const userRoot = await getUserRoot()
```

Comment:
I am not so sure about the top-level await here. Could we just run the getter when it's been used?

Reply:
`cliRoot` and `clientRoot` can be safely resolved at top-level.

About the `userRoot`, previously it was the `dirname(fullEntry)`, which didn't consider the `package.json`.

Currently `userRoot` is the "closest" package root, and `userWorkspaceRoot` is the "closest" workspace root (fallback to `userRoot` if there is no `pnpm-workspace.yaml`).

Reply:
I am mostly concerned about performance and side-effects - usually a module should not contain top-level await which will block the whole module evaluation, especially this module exposes utilties - which means every module import those utils has to wait for the functions to be executed.

Reply:
I hadn't considered that before😭. I will fix this later.